crates/vm/levm/src/vm_result.rs

```diff
@@ -0,0 +1,49 @@
+use bytes::Bytes;
+
+use crate::call_frame::Log;
+
+/// Errors that halts the program
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum VMError {
+    StackUnderflow,
+    StackOverflow,
+    InvalidJump,
+    OpcodeNotAllowedInStaticContext,
+    OpcodeNotFound,
+    InvalidBytecode,
+    FatalError, // this should never really happen
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum ResultReason {
+    Stop,
+    Return,
+}
+
+/// Result of a transaction execution.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum ExecutionResult {
+    Success {
+        reason: ResultReason,
+        logs: Vec<Log>,
+        return_data: Bytes,
+    },
+    /// Reverted by `REVERT` opcode that doesn't spend all gas.
+    Revert { gas_used: u64, output: Bytes },
+    /// Reverted for various reasons and spend all gas.
+    Halt {
```

Comment:
I'm not sure if `Halt` is the most appropriate name for this because it may not be descriptive enough. Other instructions Halt too but they don't consume all gas. Maybe `Failure` or something like that would be more accurate.
In [EVM Codes](https://www.evm.codes/) they use the word Halt just to say that an opcode stops the vm execution.

Reply:
I think `Halt` is the correct name and we should keep it, because that's the name that it is used in the yellowpaper and every doc. I think the issue might be that right now I am using `Halt` when there is an error in some opcodes, when I should be using `Revert`. Let me make some changes and I'll ping you when I'm done!

Reply:
I've added a reason to `Revert` so in the future we can tell better what caused it. Regarding the halts left I would leave them as they are, mainly because we are only halting in the `JUMP` ops, and the yellowpaper says that it should halt here (when there is an invalid jump dest).